### PR TITLE
feat(obligation): add license obligations to a project

### LINKED ISSUE #2291: Add license obligations to a project.
### Description

This new endpoint will add license obligations from license database on the basis of obligation ids.

### PR DESCRIPTION
**Pull request by rudra-superrr**
Issue: Closes #2291 

Description: Add license obligations to a project

### CODE REVIEW

**Comment by keerthi-bl** [comment]
Taken this PR for testing

**Comment by keerthi-bl** [comment]
Test successful.

image

2292_PR_2

**Comment by rudra-superrr** [comment]
Not able to update project's obligation after obligation type value added to the license obligation table.

**Comment by rudra-superrr** [comment]
Merge this PR after #2321 PR, then user will be able to update license obligation table

**Comment by keerthi-bl** [comment]
Test Successful. Working as expected.

image

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/project/Sw360ProjectService.java [hunk 1]
```diff
@@ -28,16 +29,28 @@
...
+import org.eclipse.sw360.datahandler.thrift.components.ComponentService;
 import org.eclipse.sw360.datahandler.thrift.components.Release;
 import org.eclipse.sw360.datahandler.thrift.components.ReleaseClearingStatusData;
 import org.eclipse.sw360.datahandler.thrift.components.ReleaseLink;
+import org.eclipse.sw360.datahandler.thrift.licenseinfo.LicenseInfoParsingResult;
+import org.eclipse.sw360.datahandler.thrift.licenseinfo.LicenseInfoService;
+import org.eclipse.sw360.datahandler.thrift.licenseinfo.LicenseObligationsStatusInfo;
+import org.eclipse.sw360.datahandler.thrift.licenseinfo.ObligationParsingResult;
+import org.eclipse.sw360.datahandler.thrift.licenses.LicenseService;
+import org.eclipse.sw360.datahandler.thrift.licenses.ObligationLevel;
```
Comment (hoangnt2) [inline rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/project/Sw360ProjectService.java:44]: Unused import

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/project/Sw360ProjectService.java [hunk 2]
```diff
@@ -28,16 +29,28 @@
...
+import org.eclipse.sw360.datahandler.thrift.licenses.License;
+import org.eclipse.sw360.datahandler.thrift.licenses.ObligationType;
```
Comment (hoangnt2) [inline rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/project/Sw360ProjectService.java:46]: Unused import

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/project/Sw360ProjectService.java [hunk 3]
```diff
@@ -57,14 +70,22 @@
 import org.springframework.security.access.AccessDeniedException;
 import org.springframework.stereotype.Service;
 
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Sets;
+
 import javax.annotation.PreDestroy;
+
+import java.io.IOException;
```
Comment (hoangnt2) [inline rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/project/Sw360ProjectService.java:79]: Unused import

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/project/Sw360ProjectService.java [hunk 4]
```diff
@@ -134,6 +157,299 @@ public String getCyclicLinkedProjectPath(Project project, User user) throws TExc
         return cyclicLinkedProjectPath;
     }
 
+    public Map<String, Set<Release>> getLicensesFromAttachmentUsage(
+            Map<String, AttachmentUsage> licenseInfoAttachmentUsage,
+            Map<String, ProjectReleaseRelationship> releaseIdToUsage, User user) {
```
Comment (hoangnt2) [inline rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/project/Sw360ProjectService.java:162]: Unused param **Map<String, ProjectReleaseRelationship> releaseIdToUsage**

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/project/Sw360ProjectService.java [hunk 5]
```diff
@@ -134,6 +157,299 @@ public String getCyclicLinkedProjectPath(Project project, User user) throws TExc
...
+                lic = licenseClient.getByID(entry.getKey(), user.getDepartment());
+            } catch (TException exp) {
+                log.warn("Error fetching license from backend! License Id-" + entry.getKey(), exp.getMessage());
+                return;
+            }
+            if (lic == null || CommonUtils.isNullOrEmptyCollection(lic.getObligations()))
+                return;
+
+            lic.getObligations().stream().filter(Objects::nonNull).forEach(obl -> {
+                String keyofObl = CommonUtils.isNotNullEmptyOrWhitespace(obl.getTitle()) ? obl.getTitle()
```
Comment (hoangnt2) [inline rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/project/Sw360ProjectService.java:312]:
```suggestion
                String keyOfObl = CommonUtils.isNotNullEmptyOrWhitespace(obl.getTitle()) ? obl.getTitle()
```

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/project/Sw360ProjectService.java [hunk 6]
```diff
@@ -134,6 +157,299 @@ public String getCyclicLinkedProjectPath(Project project, User user) throws TExc
...
+                    osi.setLicenseIds(licenseIds);
+                }
+                licenseIds.add(entry.getKey());
+            });
+
+        }));
+        return obligationStatusMap;
+    }
+
+    public RequestStatus updateLinkedObligations(Project project, User user, Map<String, ObligationStatusInfo> licenseObligation) throws TException {
```
Comment (hoangnt2) [inline rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/project/Sw360ProjectService.java:338]: **TException** is never thrown

Comment (rudra-superrr) [reply rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/project/Sw360ProjectService.java:335]: Used on line 367.

**Comment by rudra-superrr** [comment]
Hi @hoangnt2 , review comments addressed.

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/project/Sw360ProjectService.java (hunk 6 above)
Comment (hoangnt2) [reply rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/project/Sw360ProjectService.java:335]: @rudra-superrr, but you wrapped your code snippet in a try-catch block, so I don't think this function can throw Exception

Comment (rudra-superrr) [reply rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/project/Sw360ProjectService.java]: Understood.

**Comment by hoangnt2** [comment]
Code look good

**Pull request merged**